### ISSUE TITLE
Results not good.

### ISSUE THREAD

**Issue by farizrahman4u** [bitwise-ben/Snake#1]
@bitwise-ben 
After 10,000 epochs, I am still not able to obtain results comparable to the gif in readme. Can you provide the exact parameters you used?

**Comment by bitwise-ben** [comment]
That's unfortunate. Have you pulled the latest commits? I've used the exact same parameters as currently provided in the master branch.

**Comment by SergeevPavel** [comment]
I have the same troubles. After fitting with parameters from master loss function not decrease and agent behavior was completely wrong. Maybe it depends on some random factors like weights initialization. In my case at the beginning loss function was something about 0.1 but at last 1000 epochs it start to increase and obtain about 10000.

**Comment by bitwise-ben** [comment]
Hmm... I'll have to take a look at it. So far, the model has found a satisfactory solution in 3 out of 3 training sessions. I have also seen the loss function increase a bit towards the end (although not as much as @SergeevPavel has reported), and I assumed that was due to the annealing of the epsilon parameter.

**Comment by farizrahman4u** [comment]
Well, Its something with initialization. Its working good most of the time. Sometimes the loss just shoots up to some very big number.

**Comment by jinchenglee** [comment]
@bitwise-ben 

1. How long (in terms of hours) did it take you to train the agent to behave like the gif?
2. I don't see you take advantage of the distance between snake head toward fruit/food, do you? Have you considered to take that into account for cost function? Gut feeling that should help.

**Comment by farizrahman4u** [comment]
@jinchenglee 

> I don't see you take advantage of the distance between snake head toward fruit/food, do you? Have you considered to take that into account for cost function? Gut feeling that should help.

It will, for sure. But that would be cheating. The NN has to figure that out.

> How long (in terms of hours) did it take you to train the agent to behave like the gif?

The current implementation is not very efficient.. you could use the one at qlearning4k instead, it is much faster.

**Comment by bitwise-ben** [comment]
@jinchenglee, the model (as is) runs for at least a few hours, but the code can be optimized quite a bit by "vectorizing" the most inner "for" loop in the training procedure.

**Comment by farizrahman4u** [comment]
Yep.. thats what I did..
